### PR TITLE
feat(manager): support azure-pipelines

### LINKED ISSUE #7324: Support for azure-pipelines repository and container resources
**What would you like Renovate to be able to do?**

Similar to CircleCI, I would like to have automatic upgrade of Azure pipelines configuration.
It should upgrade repository and container resources defined in the default `azure-pipelines.yaml` files.

See yaml schema: https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema%2Cparameter-schema#resources

**Is this a feature you'd be interested in implementing yourself?**

Yes, I started implementing it, PR will come in a few minutes.

### PR DESCRIPTION
**Pull request by etiennetremel**
Add support for azure-pipelines configuration.

Closes https://github.com/renovatebot/renovate/issues/7324

### CODE REVIEW

File: lib/manager/azure-pipelines/index.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export { extractPackageFile } from './extract';
+
+export const defaultConfig = {
+  fileMatch: ['(^|/)azure-pipelines?\\.ya?ml'],
```
Comment (JamieMagee) [inline lib/manager/azure-pipelines/index.ts:4]: This is a tricky one, as you can select any `.yaml` or `.yml` file when setting up a pipeline, however you're correct that these filenames are the convention.

I would, however, add some more. I've also seen `.azure-pipelines.yaml` (and all the variations) as well as a directory called `.azure-pipelines` and any `.yaml`/`.yml` file under that directory.

File: lib/manager/azure-pipelines/readme.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+The `azure-pipelines` manager extracts container and repository resources from
+the `resources:` block. For example:
```
Comment (JamieMagee) [inline lib/manager/azure-pipelines/readme.md:2]: Manual line breaks

```suggestion
The `azure-pipelines` manager extracts container and repository resources from the `resources:` block. For example:
```

File: lib/manager/azure-pipelines/readme.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      ref: refs/tags/v0.5.1
+  containers:
+    - container: linux
+      image: ubuntu:16.04
+    - container: python
+      image: python:3.7@sha256:3870d35b962a943df72d948580fc66ceaaee1c4fbd205930f32e0f0760eb1077
+```
+
+More about the resources block can be found on the [Azure pipelines
+documentation](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema%2Cparameter-schema#resources).
```
Comment (JamieMagee) [inline lib/manager/azure-pipelines/readme.md:21]: Manual line breaks

```suggestion
More about the resources block can be found on the [Azure pipelines documentation](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema%2Cparameter-schema#resources).
```

File: lib/manager/azure-pipelines/extract.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import { logger } from '../../logger';
+import { PackageDependency, PackageFile } from '../common';
+import { getDep } from '../dockerfile/extract';
+
+interface Container {
+  image: string;
+}
+
+interface Repository {
+  type: string;
```
Comment (JamieMagee) [inline lib/manager/azure-pipelines/extract.ts:12]:
```suggestion
  type: 'git' | 'github' | 'bitbucket';
```

File: lib/manager/azure-pipelines/extract.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  return null;
+}
+
+export function parseAzurePipelines(content: string): AzurePipelines | null {
+  let pkg = null;
+  try {
+    pkg = safeLoad(content);
+  } catch (e) /* istanbul ignore next */ {
+    return null;
+  }
```
Comment (JamieMagee) [inline lib/manager/azure-pipelines/extract.ts:75]: We should log this.

File: lib/manager/azure-pipelines/extract.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  return {
+    autoReplaceStringTemplate: 'refs/tags/{{newValue}}',
+    currentValue: repository.ref.replace('refs/tags/', ''),
+    datasource: datasourceGitTags.id,
+    depName: repository.name,
+    depType: 'gitTags',
+    lookupName: `git@github.com:${repository.name}.git`,
```
Comment (JamieMagee) [inline lib/manager/azure-pipelines/extract.ts:43]: For a readonly option, I think using HTTPS might be easier.

**Comment by JamieMagee** [comment]
This is a good contribution, and not something I'd considered Renovate doing! Now that you've got me thinking, I would really love to see Renovate handle the task versions i.e. `- task: GitHubRelease@0` to `- task: GitHubRelease@1` similar to how GitHub Actions are handled.

I think the test failures are just intermittent, so I've kicked them off again.

File: lib/manager/azure-pipelines/index.ts (hunk 1 above)
Comment (rarkins) [inline lib/manager/azure-pipelines/index.ts:4]:
```suggestion
  fileMatch: ['azure.*pipelines?.*\\.ya?ml'],
```

How about this? i.e. any yml/yaml file with `azure` and `pipelines` occurring in the directory of filename?

Comment (etiennetremel) [reply lib/manager/azure-pipelines/index.ts]: Thanks, I updated the regexp with your suggestion

File: lib/manager/azure-pipelines/index.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+export { extractPackageFile } from './extract';
+
+export const defaultConfig = {
+  fileMatch: ['azure.*pipelines?.*.ya?ml$'],
```
Comment (rarkins) [inline lib/manager/azure-pipelines/index.ts:4]:
```suggestion
  fileMatch: ['azure.*pipelines?.*\\.ya?ml$'],
```

Comment (viceice) [reply lib/manager/azure-pipelines/index.ts:4]: What about a starting optional dot?

Comment (rarkins) [reply lib/manager/azure-pipelines/index.ts:4]: There's no `^` so the above will allow dot and more. e.g. could even be `config/azure-pipeline.yml`

Comment (viceice) [reply lib/manager/azure-pipelines/index.ts:4]: Yes, you are right. Have not seen it. 🙈

**Review by JamieMagee — DISMISSED** [review]

**Comment by rarkins** [comment]
Missing coverage on extract.ts lines 66 and 114

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs coverage fix

**Review by JamieMagee — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**